### PR TITLE
C04hcp traversal

### PR DESCRIPTION
**Pull request by SabrinaKrallmann**
# Description

Create c04HCP Traversal accordingly to "Algorithmic and Implementational Optimizations of Molecular Dynamics Simulations for Process Engineering" chapter 8: outlook. And comment on c04Traversal.

# How Has This Been Tested?

- [x] TraversalTest

### CODE REVIEW

File: src/autopas/containers/linkedCells/traversals/C04HCP.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+            break;
+          case 2:
+            x_index -= 2;
+            break;
+        }
+        if ((y - startY) % 2 != 0) {  // shift x-axis every second y-row
+          x_index += 3;
+        }
+        const std::array<long, 3> base3DIndex = {x_index, y, z};
+        processBasePack6(cells, base3DIndex);
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C04HCP.h:192]:
```suggestion
        processBasePack6(cells, {x_index, y, z});
```
no need for temporary

File: src/autopas/containers/linkedCells/traversals/C04HCP.h [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      for (long x = startX; x < (endX + 4);
+           x += 6) {  // color starts every 6th column again, the +4 is needed to prevent ending too early, since it
+                      // will be shifted back inside the loop
+        long x_index = x;
+        // shift on x-axis according to z-value
+        // first: no shift
+        // second: -4 shift
+        // third: -2 shift
+        // fourth: go back to first
+        switch ((z - startZ) % 12 / 4) {
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C04HCP.h:178]: explain where the 12 comes from

File: src/autopas/containers/linkedCells/traversals/C04HCP.h [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      break;
+    case 3:
+      startOfThisColor = {-2l, 0l, -1l};
+      break;
+  }
+
+  // to fix compiler complaints about perfectly nested loop.
+  long startX = startOfThisColor[0], endX = _end[0];
+  long startY = startOfThisColor[1], endY = _end[1];
+  long startZ = startOfThisColor[2], endZ = _end[2];
```
[the remark this answers is not in the document]
Comment (SabrinaKrallmann) [reply src/autopas/containers/linkedCells/traversals/C04HCP.h:160]: This is actually a comment, that I copied from the c04 traversal sometime in the beginning along with the initialization. I did not test it myself, if the compiler complains.
Some of the variables change, some remain, so I will mark those who remain const.

File: src/autopas/containers/linkedCells/traversals/C04Traversal.h [hunk 4]
```diff
@@ -219,7 +219,7 @@ void C04Traversal<ParticleCell, PairwiseFunctor, dataLayout, useNewton3>::traver
     correctParity += 4;
   }
 
-  // to fix compiler complaints about perfectly nested loop.
+  // to fix intel64 icpc compiler complaints about perfectly nested loop.
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C04Traversal.h:222]:
```suggestion
  // to fix intel64 icpc compiler complaints about perfectly nested loop (tested with version 19.0.4.20190416).
```

**Pull request merged**